File: 1016-subarray-sums-divisible-by-k/1016-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        unordered_map<int, int> mp; mp[0] = 1;

        int count = 0, sum = 0;
        for(int num: nums){
            sum += num;
            int rem = sum % k;
            if(rem < 0) rem += k;
            if(mp[rem]) 
                count += mp[rem];
            mp[rem]++;
        }
        return count;
    }
};

File: 1016-subarray-sums-divisible-by-k/1016-subarray-sums-divisible-by-k.py
class Solution(object):
    def subarraysDivByK(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        # dictonery
        mp = {0: 1}
        count, prefix = 0, 0

        for num in nums:
            prefix += num
            remainder = prefix % k
            if remainder in mp:
                count += mp[remainder]
            
            # if the remainder doesn't exist then return 0 them increment by 1
            mp[remainder] = mp.get(remainder, 0) + 1
        
        return count

        